### DISCUSSION TITLE
What other resources?

### DISCUSSION THREAD

**Discussion by wz2b** [onhate/cdk-gc#1] [General]
Following [the original issue](https://github.com/aws/aws-cdk-rfcs/issues/64) that inspired this, they've admitted they are never going to do anything about it.  That means we should make this tool as strong and complete as possible.

I'm wondering about [this line](https://github.com/onhate/cdk-gc/blob/504a0f894d36edff4230aaed7666987f1e18a582/src/gc.js#L125) though.  The only assets you look for are lambdas.  That's mainly what I use as well, but I'm wondering what other assets could be in there that might need cleanup as well - for example if someone is using [@aws-cdk/aws-s3-deployment](https://docs.aws.amazon.com/cdk/api/v1/docs/aws-s3-deployment-readme.html) does that end up the staging bucket as well?  What other things could end up in the staging bucket?  The thing that troubles me is I don't use every single aws service that could be instantiated by either cdk or cloudformation itself, so I'm not very confident that lambdas are the only thing.

[Assets are local files, directories, or Docker images that can be bundled into AWS CDK libraries and apps. For example, an asset might be a directory that contains the handler code for an AWS Lambda function. Assets can represent any artifact that the app needs to operate.](https://docs.aws.amazon.com/cdk/v2/guide/assets.html)

**Comment by onhate** [comment]
I'm in the same scenario, I only know how to detect used assets for lambda because they are referred in the stack, another one we could work on is the docker images on ECR, but other than that I'm not sure how it works.

**Reply by wz2b** [reply]
I was thinking about this some more.  I'm not sure this is an actionable idea but anything that references an asset in the bucket must refer to that bucket somehow.  So in the case of a lambda, I get:

```
    "Code": {
     "S3Bucket": "cdk-hnb659fds-assets-xxxxxxxxxxxxx-us-east-1",
     "S3Key": "e2afe2444e8ec790daead3326fefdb20b1293cxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.zip"
    },
```

so maybe the 'universal' way to do this is to just look for some reference to the bucket.  It could be in the form shown above (for lambdas) but I get the sense it might also be something like "s3://cdk-hnb659fds-assets-xxxxxxxxxxxxx-us-east-1/blahblahblah" 

I'm thinking a good way to start might be to just do a dumb search through the templates and find any string that refers to that bucket name and issue a warning.  That might be a little noisy in some cases though, for example I actually generate some buckets in my S3 then store the names as stack parameters so other stacks can find them.  That would cause a false alarm.

Really, the ideal thing would be if we could figure out more definitively what other kinds of assets cdk puts in there.  I just have zero real world experience with any assets other than lambda code.
